File: android/vcam-bridge-app/src/dev/vcam/bridge/MainActivity.java
package dev.vcam.bridge;

import android.Manifest;
import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

public final class MainActivity extends Activity {
    private static final int CAMERA_PERMISSION_REQUEST = 100;
    private TextView statusView;

    private final BroadcastReceiver statusReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String status = intent.getStringExtra(CameraBridgeService.EXTRA_STATUS);
            if (status != null) {
                statusView.setText(status);
            }
        }
    };

    @Override
    protected void onCreate(Bundle state) {
        super.onCreate(state);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(40, 72, 40, 40);

        TextView title = new TextView(this);
        title.setText("Deep Live Camera Bridge");
        title.setTextSize(24.0f);
        layout.addView(title);

        statusView = new TextView(this);
        statusView.setText("Camera2 → GPU orientation → hardware H.264 → Windows 192.168.1.35\n"
                + "Processed return appears as camera 120. Orientation defaults to Auto and can "
                + "be changed in the Arch control center.");
        statusView.setTextSize(16.0f);
        statusView.setPadding(0, 32, 0, 32);
        layout.addView(statusView);

        Button start = new Button(this);
        start.setText("Start bridge");
        start.setOnClickListener(view -> ensurePermissionAndStart());
        layout.addView(start);

        Button stop = new Button(this);
        stop.setText("Stop bridge");
        stop.setOnClickListener(view -> {
            Intent intent = new Intent(this, CameraBridgeService.class);
            intent.setAction(CameraBridgeService.ACTION_STOP);
            startService(intent);
            statusView.setText("Bridge stopped; camera 120 remains on its fallback frame.");
        });
        layout.addView(stop);

        setContentView(layout);
        ensurePermissionAndStart();
    }

    private void ensurePermissionAndStart() {
        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.CAMERA}, CAMERA_PERMISSION_REQUEST);
            return;
        }
        if (Build.VERSION.SDK_INT >= 33
                && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS}, 101);
        }
        Intent intent = new Intent(this, CameraBridgeService.class);
        intent.setAction(CameraBridgeService.ACTION_START);
        if (Build.VERSION.SDK_INT >= 26) {
            startForegroundService(intent);
        } else {
            startService(intent);
        }
        statusView.setText("Starting physical-camera encoder…");
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] results) {
        super.onRequestPermissionsResult(requestCode, permissions, results);
        if (requestCode == CAMERA_PERMISSION_REQUEST) {
            if (results.length > 0 && results[0] == PackageManager.PERMISSION_GRANTED) {
                ensurePermissionAndStart();
            } else {
                statusView.setText("Camera permission is required for the bridge input.");
            }
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        IntentFilter filter = new IntentFilter(CameraBridgeService.ACTION_STATUS);
        if (Build.VERSION.SDK_INT >= 33) {
            registerReceiver(statusReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
        } else {
            registerReceiver(statusReceiver, filter);
        }
    }

    @Override
    protected void onStop() {
        unregisterReceiver(statusReceiver);
        super.onStop();
    }
}

File: android/vcam-bridge-app/src/dev/vcam/bridge/MainActivity.kt
package dev.vcam.bridge

import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat

/**
 * Main UI: bridge start/stop + camera list + live preview toggle.
 * Merges the old bridge app and camera2-test app into one UI.
 */
class MainActivity : AppCompatActivity() {

    private lateinit var statusView: TextView
    private lateinit var logView: TextView
    private lateinit var camListView: TextView
    private lateinit var cameraManager: CameraManager
    private val logLines = ArrayDeque<String>()
    private val maxLogLines = 300

    private val statusReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val status = intent.getStringExtra(CameraBridgeService.EXTRA_STATUS)
            if (status != null) log("bridge: $status")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        buildUi()
        cameraManager = getSystemService(CAMERA_SERVICE) as CameraManager
        checkPermissionsAndStart()
    }

    private fun buildUi() {
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(24, 48, 24, 24)
            setBackgroundColor(0xFF0A0E14.toInt())
        }

        val title = TextView(this).apply {
            text = "VCam Bridge"
            textSize = 22f
            setTextColor(0xFF7EE787.toInt())
        }
        root.addView(title)

        statusView = TextView(this).apply {
            textSize = 14f
            setTextColor(0xFFD7E0EA.toInt())
            setPadding(8, 16, 8, 12)
            text = "Ready"
        }
        root.addView(statusView)

        val btnRow = LinearLayout(this).apply { orientation = LinearLayout.HORIZONTAL }
        btnRow.addView(Button(this).apply {
            text = "Start Bridge"
            setOnClickListener { startBridge() }
        }, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))
        btnRow.addView(Button(this).apply {
            text = "Stop Bridge"
            setOnClickListener { stopBridge() }
        }, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))
        btnRow.addView(Button(this).apply {
            text = "Preview"
            setOnClickListener { startActivity(Intent(this@MainActivity, PreviewActivity::class.java)) }
        }, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))
        root.addView(btnRow)

        val camLabel = TextView(this).apply {
            textSize = 13f
            setTextColor(0xFF9DACBB.toInt())
            setPadding(0, 12, 0, 4)
            text = "Cameras:"
        }
        root.addView(camLabel)
        camListView = TextView(this).apply {
            textSize = 12f
            setTextColor(0xFF8B949E.toInt())
            setPadding(0, 0, 0, 12)
        }
        root.addView(camListView)
        refreshCameraList()

        val refreshBtn = Button(this).apply {
            text = "Refresh Cameras"
            setOnClickListener { refreshCameraList() }
        }
        root.addView(refreshBtn)

        logView = TextView(this).apply {
            textSize = 11f
            setTextColor(0xFF79C0FF.toInt())
            setPadding(8, 8, 8, 8)
            setBackgroundColor(0xFF161B22.toInt())
        }
        root.addView(ScrollView(this).apply { addView(logView) },
            LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f))

        setContentView(root)
    }

    private fun refreshCameraList() {
        try {
            val sb = StringBuilder()
            for (id in cameraManager.cameraIdList) {
                val cc = cameraManager.getCameraCharacteristics(id)
                val facing = when (cc.get(CameraCharacteristics.LENS_FACING)) {
                    CameraCharacteristics.LENS_FACING_FRONT -> "front"
                    CameraCharacteristics.LENS_FACING_BACK -> "back"
                    CameraCharacteristics.LENS_FACING_EXTERNAL -> "external"
                    else -> "?"
                }
                val level = when (cc.get(CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL)) {
                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_3 -> "level_3"
                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_FULL -> "full"
                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LIMITED -> "limited"
                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LEGACY -> "legacy"
                    else -> "?"
                }
                val orient = cc.get(CameraCharacteristics.SENSOR_ORIENTATION)
                sb.appendLine("  $id: facing=$facing level=$level orient=$orient")
            }
            camListView.text = sb.toString()
        } catch (e: Exception) {
            camListView.text = "Error: ${e.message}"
        }
    }

    private fun startBridge() {
        val intent = Intent(this, CameraBridgeService::class.java).apply {
            action = CameraBridgeService.ACTION_START
        }
        if (Build.VERSION.SDK_INT >= 26) startForegroundService(intent) else startService(intent)
        log("bridge: starting...")
    }

    private fun stopBridge() {
        startService(Intent(this, CameraBridgeService::class.java).apply {
            action = CameraBridgeService.ACTION_STOP
        })
        log("bridge: stopped")
    }

    private fun log(msg: String) {
        logLines.addLast(msg)
        while (logLines.size > maxLogLines) logLines.removeFirst()
        logView.text = logLines.joinToString("\n")
    }

    private fun checkPermissionsAndStart() {
        val perms = mutableListOf(Manifest.permission.CAMERA)
        if (Build.VERSION.SDK_INT >= 33) perms.add(Manifest.permission.POST_NOTIFICATIONS)
        val missing = perms.filter {
            ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
        }
        if (missing.isNotEmpty()) {
            requestPermissions(missing.toTypedArray(), 100)
        } else {
            startBridge()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<out String>, results: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, results)
        if (requestCode == 100 && results.isNotEmpty()
            && results[0] == PackageManager.PERMISSION_GRANTED) {
            startBridge()
        }
    }

    override fun onStart() {
        super.onStart()
        val filter = IntentFilter(CameraBridgeService.ACTION_STATUS)
        if (Build.VERSION.SDK_INT >= 33) {
            registerReceiver(statusReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
        } else {
            registerReceiver(statusReceiver, filter)
        }
    }

    override fun onStop() {
        unregisterReceiver(statusReceiver)
        super.onStop()
    }
}
